miden-lib/asm/miden/kernels/tx/tx.masm

```diff
@@ -85,71 +89,180 @@ proc.increment_num_created_notes
-    # => [tag_low, note_type, ASSET, tag, note_type, RECIPIENT]
+    # => [tag_low, note_type, tag, note_type, RECIPIENT]
 
     u32shr.30 u32and assertz.err=ERR_NOTE_INVALID_TAG_PREFIX_FOR_TYPE
-    # => [ASSET, tag, note_type, RECIPIENT]
+    # => [tag, aux, note_type, RECIPIENT]
 
     # get the index for the next note to be created and increment counter
     exec.increment_num_created_notes
-    # => [note_idx, ASSET, tag, note_type, RECIPIENT]
+    # => [note_idx, tag, aux, note_type, RECIPIENT]
 
     # get a pointer to the memory address at which the note will be stored
     exec.memory::get_created_note_ptr
-    # => [note_ptr, ASSET, tag, note_type, RECIPIENT]
+    # => [note_ptr, tag, aux, note_type, RECIPIENT]
 
     # populate the metadata
-    movup.5 exec.account::get_id movup.7
-    # => [note_type, sender_acct_id, tag, note_ptr, ASSET, RECIPIENT]
-
-    push.0 # TODO: allow the user to push the aux data
-    # => [aux, note_type, sender_acct_id, tag, note_ptr, ASSET, RECIPIENT]
-
+    movdn.3 movdn.2 exec.account::get_id movdn.2
+    # => [aux, note_type, sender_acct_id, tag, note_ptr, RECIPIENT]
+    
     emit.NEW_NOTE_EVENT
 
     # set the metadata for the new created note
     dup.4 exec.memory::set_created_note_metadata
-    # => [note_ptr, ASSET, RECIPIENT]
+    # => [note_ptr, RECIPIENT]
 
-    # set the number of assets for the note to 1
-    push.1 dup.1 exec.memory::set_created_note_num_assets
-    # => [note_ptr, ASSET, RECIPIENT]
+    # set the RECIPIENT
+    dup movdn.5 exec.memory::set_created_note_recipient
+    # => [note_ptr]
 
-    movdn.4 padw swapw movup.8
-    # => [note_ptr, ASSET, 0, 0, 0, 0, RECIPIENT]
+    # set the number of assets to 0
+    dup push.0 swap exec.memory::set_created_note_num_assets
+    # => [note_ptr]
+end
 
-    # add the asset to the note
-    dup movdn.5 exec.memory::get_created_note_asset_data_ptr mem_storew dropw
-    # => [note_ptr, 0, 0, 0, 0, RECIPIENT]
+#! Adds an ASSET to the specified note and returns a pointer to the memory address
+#! at which the note is stored.
+#!
+#! Inputs: [ptr, ASSET]
+#! Outputs: [ptr]
+#!
+#! ptr is the pointer to the memory address at which the note is stored.
+#! ASSET can be a fungible or non-fungible asset.
+export.move_asset_to_note
+    # validate the ASSET
+    movdn.4 exec.asset::validate_asset
+    # => [ASSET, ptr]
 
-    movdn.8 swapw padw swapw movup.12
-    # => [note_ptr, RECIPIENT, 0, 0, 0, 0, 0, 0, 0, 0]
+    # check if the note has at least one ASSET already
+    dup.4 exec.memory::get_created_note_num_assets dup movdn.6 neq.0
+    # => [num_of_assets>0?, ASSET, ptr, num_of_assets]
+
+    if.true
+        exec.check_if_asset_already_exists
+        # => [asset_ptr, ASSET'', ptr, num_of_assets]
+    else
+        # There is no asset in the note yet, so we add the asset at the first position
+        dup.4 exec.memory::get_created_note_asset_data_ptr
+        # => [asset_ptr, ASSET, ptr, num_of_assets]
+    end
```

Comment:
I would have probably taken a slightly different approach here:

- First, we could validate the asset (as you have now) and determine its type (fungible or non-fungible).
- Then, for non-fungible assets, I'd create a helper procedure - e.g., `add_non_fungible_asset_to_note`. In this function, we could use simple `eqw` comparisons in a loop, and then, at the end of the loop, if nothing failed, add the asset to the note. So, the function would be relatively simple.
- For fungible assets, I'd create a separate helper procedure - e.g., `add_fungible_asset_to_note`. I'd also create two other helper procedures:
  - Something like `asset::get_issuer` which would return the faucet ID that issued the asset.
  - Something like `asset::combine_fungible_assets` which would combine two fungible assets that have the same origin.
  - I would then use these functions similar to how you have now.

While this will probably result in a bit more code - I think the code would be more encapsulated and easier to follow/maintain (e.g., if the structure of assets changes, we would only need to update the relevant helper functions).